src/workerd/io/io-context.h

```diff
@@ -1257,23 +1257,40 @@ kj::_::ReducePromises<RemoveIoOwn<T>> IoContext::awaitJs(jsg::Lock& js, jsg::Pro
   auto paf = kj::newPromiseAndFulfiller<RemoveIoOwn<T>>();
   struct RefcountedFulfiller: public Finalizeable, public kj::Refcounted {
     kj::Own<kj::PromiseFulfiller<RemoveIoOwn<T>>> fulfiller;
+    kj::Own<const workerd::Worker::Isolate> isolate;
     bool isDone = false;
 
-    RefcountedFulfiller(kj::Own<kj::PromiseFulfiller<RemoveIoOwn<T>>> fulfiller)
-        : fulfiller(kj::mv(fulfiller)) {}
+    RefcountedFulfiller(kj::Own<const workerd::Worker::Isolate> isolate,
+        kj::Own<kj::PromiseFulfiller<RemoveIoOwn<T>>> fulfiller)
+        : fulfiller(kj::mv(fulfiller)),
+          isolate(kj::mv(isolate)) {}
 
     ~RefcountedFulfiller() noexcept(false) {
       if (!isDone) {
-        // The JavaScript resolver was garbage collected, i.e. JavaScript will never resolve
-        // this promise.
-        fulfiller->reject(JSG_KJ_EXCEPTION(FAILED, Error, "Promise will never complete."));
+        auto hasExcessivelyExceededHeapLimit =
+            isolate->getLimitEnforcer().hasExcessivelyExceededHeapLimit();
```

Comment:
hmm.. this *should* be ok I think? But does make me somewhat nervous. The references to the `RefcountedFulfiller` *should* be dropped before the isolate is destroyed but we should just double check and confirm that having the RefcountedFulfiller hold a strong reference to the isolate won't cause any problems. I'm probably being needlessly paranoid here tho.

Reply:
Yeah, I've been meaning to look into this too. Is there any other way I could access the IsolateLimitEnforcer here?

Reply:
If the isolate was already dropped returning a jsg.Error has no meaning.
Why not hold a WeakRef and only run the hasExcessivelyExceededHeapLimit specialization if we're able to get a stong ref?

Reply:
isolate.getWeakRef() is an easy way to get a weak ref btw.

Reply:
Nice idea, done!